src/main/scala/org/graphframes/pattern/patterns.scala

```diff
@@ -48,13 +49,124 @@ private[graphframes] object PatternParser extends RegexParsers {
         result.asInstanceOf[Success[Seq[Pattern]]].get
       case result: NoSuccess =>
         throw new InvalidParseException(
           s"Failed to parse bad motif string: '$s'.  Returned message: ${result.msg}")
     }
+    assertValidPatterns(result)
+    result
+  }
+
+  /**
+   * Checks all Patterns for validity:
+   *  - Disallow named edges within negated terms
+   *  - Disallow term "()-[]->()" and its negation
+   *  - Disallow name to be shared by a vertex and an edge
+   * @throws InvalidParseException if an negated terms contain named edges
+   */
+  private def assertValidPatterns(patterns: Seq[Pattern]): Unit = {
+
+    // vertexNames, edgeNames are used to check for duplicate names across vertices and edges
+    val vertexNames = mutable.HashSet.empty[String]
+    val edgeNames = mutable.HashSet.empty[String]
+    def addVertex(v: Vertex): Unit = v match {
+      case NamedVertex(name) =>
+        if (edgeNames.contains(name)) {
+          throw new InvalidParseException(s"Motif reused name '$name' for both a vertex and " +
+            s"an edge, which is not allowed.")
+        }
+        vertexNames += name
+      case AnonymousVertex =>  // pass
+    }
+    def addEdge(e: Edge): Unit = e match {
+      case NamedEdge(name, src, dst) =>
+        if (vertexNames.contains(name)) {
+          throw new InvalidParseException(s"Motif reused name '$name' for both a vertex and " +
+            s"an edge, which is not allowed.")
+        }
+        edgeNames += name
+        addVertex(src)
+        addVertex(dst)
+      case AnonymousEdge(src, dst) =>
+        addVertex(src)
+        addVertex(dst)
+    }
+
+    patterns.foreach {
+      case Negation(edge) =>
+        edge match {
+          case NamedEdge(name, src, dst) =>
+            throw new InvalidParseException(s"Motif finding does not support negated named " +
+              s"edges, but the given pattern contained: !($src)-[$name]->($dst)")
+          case AnonymousEdge(AnonymousVertex, AnonymousVertex) =>
+            throw new InvalidParseException(s"Motif finding does not support completely " +
+              s"anonymous negated edges !()-[]->().  Users can check for 0 edges in the graph " +
+              s"using the edges DataFrame.")
+          case e @ AnonymousEdge(_, _) =>
+            addEdge(e)
+        }
+      case AnonymousEdge(AnonymousVertex, AnonymousVertex) =>
+        throw new InvalidParseException(s"Motif finding does not support completely " +
+          s"anonymous edges ()-[]->().  Users can check for the existence of edges in the " +
+          s"graph using the edges DataFrame.")
+      case e @ AnonymousEdge(_, _) =>
+        addEdge(e)
+      case e @ NamedEdge(_, _, _) =>
+        addEdge(e)
+      case AnonymousVertex => // ok
+      case v @ NamedVertex(_) =>
+        addVertex(v)
+    }
+  }
+
+  /**
+   * Split patterns into positive (non-negated) and negated terms, and find all named vertices
+   * in each.
```

Comment:
The doc says "find all named vertices in each." This is a little misleading because Java/Scala takes the very first sentence as the method summary, but it is incomplete here.